Language/Haskell/GhcMod/ImportedFrom.hs

```diff
@@ -0,0 +1,826 @@
+                    Nothing -> Nothing
+
+    -- | Call @cabal sandbox hc-pkg@ to find the package the provides a module.
+    -- hcPkgFindModule :: String -> IO (Maybe String)
+    hcPkgFindModule rp m = do
+        let opts = ["sandbox", "hc-pkg", "find-module", m, "--", "--simple-output"]
+
+        x <- runCmd rp "cabal" opts
+
+        -- gmLog GmDebug "" $ strDoc $ "hcPkgFindModule stdout: " ++ show output
+        -- gmLog GmDebug "" $ strDoc $ "hcPkgFindModule stderr: " ++ show err
+        return $ case x of
+                    Just x' -> join $ (Safe.lastMay . words) <$> (Safe.lastMay . lines) x'
+                    Nothing -> Nothing
+
+    -- | Call @stack exec ghc-pkg@ to find the package the provides a module.
+    -- stackGhcPkgFindModule :: String -> IO (Maybe String)
+    stackGhcPkgFindModule rp m = do
+        let opts = ["exec", "ghc-pkg", "find-module", m, "--", "--simple-output"]
+
+        x <- runCmd rp "stack" opts
+
+        -- gmLog GmDebug "" $ strDoc $ "stackGhcPkgFindModule stdout: " ++ show output
+        -- gmLog GmDebug "" $ strDoc $ "stackGhcPkgFindModule stderr: " ++ show err
+        return $ case x of
+                    Just x' -> join $ (Safe.lastMay . words) <$> (Safe.lastMay . lines) x'
+                    Nothing -> Nothing
+
+ghcPkgHaddockUrl
+    :: forall m. (GmLog m, GmOut m, MonadIO m)
+    => FilePath
+    -> (FilePath -> [String] -> String -> IO String)
+    -> [GhcPkgDb]
+    -> String
+    -> m (Maybe String)
+ghcPkgHaddockUrl ghcPkg readProc pkgDbStack p = do
+    gmLog GmDebug "ghcPkgHaddockUrl" $ strDoc p
+
+    let p' = case splitOn "@" p of
+                [p0, _] -> p0
+                _       -> p
+
+    hout <- liftIO $ readProc ghcPkg (toDocDirOpts p' pkgDbStack) ""
+    return $ Safe.lastMay $ words $ reverse . dropWhile (== '\n') . reverse $ hout
+  where
+    -- This fails unless we have --global and --user, unlike
+    -- pkgDoc elsewhere in ghc-mod.
+    toDocDirOpts pkg dbs = ["field", pkg, "haddock-html", "--global", "--user"] ++ ghcPkgDbStackOpts dbs
+
+ghcPkgHaddockInterface
+    :: forall (m :: * -> *).  MonadIO m
+    => FilePath
+    -> (FilePath -> [String] -> String -> IO String)
+    -> [GhcPkgDb]
+    -> String
+    -> m (Maybe String)
+ghcPkgHaddockInterface ghcPkg readProc pkgDbStack p = do
+    hout <- liftIO $ readProc ghcPkg (toHaskellInterfaces p pkgDbStack) ""
+    return $ Safe.lastMay $ words $ reverse . dropWhile (== '\n') . reverse $ hout
+  where
+    toHaskellInterfaces pkg dbs = ["field", pkg, "haddock-interfaces", "--global", "--user"] ++ ghcPkgDbStackOpts dbs
+
+getVisibleExports
+    :: forall m. (GhcMonad m, MonadIO m, GmOut m, GmLog m)
+    => (String -> m (Maybe String))
+    -> String
+    -> m (Maybe (M.Map String [String]))
+getVisibleExports getHaddockInterfaces p = do
+    gmLog GmDebug "getVisibleExports" $ strDoc p
+
+    let p' = case splitOn "@" p of
```

Comment:
Seems like this should be a function, you have an identical block in the function above.